ops/testing.py

```diff
@@ -1050,6 +1052,21 @@ def __init__(self, unit_name, meta):
         # socket_path = '/charm/containers/{container_name}/pebble.socket'
         self._pebble_clients = {}  # type: {str: _TestingPebbleClient}
         self._planned_units = None
+        self._hook_is_running = ''
+
+    def _validate_relation_access(self, relation_name, relations):
+        """Ensures that the named relation exists/has been added."""
+        if len(relations) > 0:
+            return
+
+        relations = list(self._meta.peers.keys())
+        relations.extend(self._meta.requires.keys())
+        relations.extend(self._meta.provides.keys())
+        if self._hook_is_running == 'leader_elected' and relation_name in relations:
+            raise RuntimeError(
+                'cannot access relation data without first adding the relation: '
+                'use Harness.add_relation({!r}, <app>) before calling set_leader'
+                .format(relation_name))
```

Comment:
While the call to this routine's stub in production doesn't feel great, I really like this Error. It's accurate and informative, and it points charm authors to an easy fix for their tests. +1

Reply:
Maybe we should add a "You're welcome." into the error message :P